third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,193 @@
+    /// If label has multiple sources, returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    /// otherwise returns `None`
+    fn split_critical_edge(
+        &mut self,
+        attr_id: AttrId,
+        label: Label,
+    ) -> Option<(Label, Vec<Bytecode>)> {
+        // label is in `srcs_count` by construction
+        if *self.srcs_count.get(&label).expect("srcs count") > 1 {
+            Some(self.split_edge(attr_id, label))
+        } else {
+            None
+        }
+    }
+
+    /// Returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    fn split_edge(&mut self, attr_id: AttrId, label: Label) -> (Label, Vec<Bytecode>) {
+        let new_label = self.gen_fresh_label();
+        let code = vec![
+            Bytecode::Label(attr_id, new_label),
+            Bytecode::Jump(attr_id, label),
+        ];
+        (new_label, code)
+    }
+
+    /// Generates a fresh label
+    fn gen_fresh_label(&mut self) -> Label {
+        let new_label = Label::new(
+            if self.labels.is_empty() {
+                0
+            } else {
+                self.labels.iter().next_back().expect("label").as_usize() + 1
+            },
+        );
+        self.labels.insert(new_label);
+        new_label
+    }
+
+    fn emit(&mut self, bytecode: Bytecode) {
+        self.data.code.push(bytecode)
+    }
+
+    fn emit_codes(&mut self, code: Vec<Bytecode>) {
+        for instr in code {
+            self.emit(instr)
+        }
+    }
+}
+
+/// If key present in `map`, add 1 to its value; else insert 1
+fn map_inc<Key: Ord>(map: &mut BTreeMap<Key, usize>, key: Key) {
+    map.entry(key)
+        .and_modify(|n: &mut usize| {
+            let (n_suc, overflows) = n.overflowing_add(1);
+            if overflows {
+                panic!("`count_srcs` overflows")
+            } else {
+                *n = n_suc;
+            }
+        })
+        .or_insert(1usize);
+}
+
+/// Count the number of sources of labels
+/// labels with no sources are not included
+fn count_srcs(code: &[Bytecode]) -> BTreeMap<Label, usize> {
+    let mut srcs_count = BTreeMap::new();
+    for (code_offset, instr) in code.iter().enumerate() {
+        match instr {
+            Bytecode::Jump(_, label) => map_inc(&mut srcs_count, *label),
+            Bytecode::Branch(_, l0, l1, _) => {
+                map_inc(&mut srcs_count, *l0);
+                map_inc(&mut srcs_count, *l1);
+            },
+            Bytecode::Label(_, label) => {
+                if code_offset != 0 {
+                    let prev_instr = code.get(code_offset - 1).unwrap();
+                    // treat fall-through's to the label
+                    if !prev_instr.is_branch() {
+                        map_inc(&mut srcs_count, *label)
+                    }
+                }
+            },
+            _ => {},
```

Comment:
I added a prerequisite that no abort actions are used. Currently, no abort actions are used, and the transformation assumes that the only instruction with multiple successors is the branch instruction.

Reply:
Also add a comment to https://github.com/aptos-labs/aptos-core/issues/11958 to mention the dependence on that prerequisite in this file, so that it will be noticed if the assumption changes in the future.

Reply:
Still waiting for action on this.

Reply:
Oops, missed this one. Added https://github.com/aptos-labs/aptos-core/issues/11958#issuecomment-1947828209